stan/math/rev/fun/trace_quad_form.hpp

```diff
@@ -161,26 +108,25 @@ inline var trace_quad_form(const Mat1& A, const Mat2& B) {
  */
 template <typename Mat1, typename Mat2,
           require_all_matrix_t<Mat1, Mat2>* = nullptr,
-          require_any_var_matrix_t<Mat1, Mat2>* = nullptr,
-	  require_all_not_eigen_st<is_var, Mat1, Mat2>* = nullptr>
+          require_any_var_matrix_t<Mat1, Mat2>* = nullptr>
 inline var trace_quad_form(const Mat1& A, const Mat2& B) {
   check_square("trace_quad_form", "A", A);
   check_multiplicable("trace_quad_form", "A", A, "B", B);
 
-  arena_t<Mat1> arena_A = A;
-  arena_t<Mat2> arena_B = B;
+  auto arena_A = to_arena(to_var_value_if<!is_constant<Mat1>::value>(A));
```

Comment:
I ended up going with `to_var_value_if` as the lazy support for mixed `var<mat>` and `mat<var>`.

The other thing I tried was templating out the functions but that is confusing in its own way. It looks like:

```cpp
template <typename Mat1, typename Mat2,
          require_var_matrix_t<Mat1>* = nullptr,
          require_eigen_st<is_var, Mat2>* = nullptr>
inline var trace_quad_form(const Mat1& A, const Mat2& B) {
  return trace_quad_form(A, to_var_value(B));
}

template <typename Mat1, typename Mat2,
          require_eigen_st<is_var, Mat1>* = nullptr,
          require_var_matrix_t<Mat2>* = nullptr>
inline var trace_quad_form(const Mat1& A, const Mat2& B) {
  return trace_quad_form(to_var_value(A), B);
}

template <typename Mat1, typename Mat2,
          require_all_matrix_t<Mat1, Mat2>* = nullptr,	          require_all_matrix_t<Mat1, Mat2>* = nullptr,
          require_any_var_matrix_t<Mat1, Mat2>* = nullptr>	          require_any_var_matrix_t<Mat1, Mat2>* = nullptr,
	  require_all_not_eigen_st<is_var, Mat1, Mat2>* = nullptr>
inline var trace_quad_form(const Mat1& A, const Mat2& B) {
 ...
}
```

Reply:
I'm not following why they have to both be coerced to var matrices?

Reply:
The other options are:

1. Do not support mixes of `mat<var>` and `var<mat>`. I do not like that because we will have programs with both types (and so we could end up with mixed arguments)

2. Write three different specializations. I do not like that because we found out optimizing for `mat<var>` is trickier than `var<mat>`. I do not think this is particularly necessary cuz this will be faster than not having autodiff.

Reply:
Or just don't move the `matrix<var>` to a `var<matrix>`. I think Tadej and I both prefer just leaving them instead of doing the conversion. In another comment I have branch with one `quad_form` that handles all the cases of vars with one function

https://github.com/stan-dev/math/compare/feature/quad_form...feature/quad_form-pullout-logic

Reply:
I'm going to sit down one day and figure out this whole `val()` and `val_op()` thing. Something something const correctness

Reply:
The furthest I got was that it's something to do with Eigen's ```sizeof``` logic, there's a little more discussion in [this stackoverflow thread](https://stackoverflow.com/questions/56635122/using-coeffref-with-const-cwiseunaryview-failing-when-multiplying-two-cwiseuna#comment99991183_56661442). The error doesn't occur with ```float```, only with ```double```. I found a modification that fixed the compilation, but hard to know whether that would affect other behaviour. Tricky.

Reply:
> In another comment I have branch with one quad_form that handles all the cases of vars with one function

I was trying to avoid touching the pure `mat<var>` implementation since we decided to wait on benchmarks before we did that (which is why we decided not to merge https://github.com/stan-dev/math/pull/2019)

> I think Tadej and I both prefer just leaving them

Yeah I saw the comments and I made a response where I argued the case that the conversion here was better: https://github.com/stan-dev/math/pull/2164#discussion_r513022661

Since I didn't see a response to that, I just went with the conversion to move this along.

I'm not on the same page with stuff like:
```
arena_t<promote_scalar_t<var, Mat1>> arena_A(A);
arena_t<promote_scalar_t<var, Mat2>> arena_B(B);
return_t res(arena_B.val_op().transpose() * arena_A.val_op() * arena_B.val_op());
```

Like up until this point I've been assuming if we read the `.val()` of a `mat<var>` more than once, we should make a copy of it on the arena.

Since this function does a lot of `.val()` calls on `A` and `B`, my understanding is that we need to make arena copies of the values, otherwise we're going to pointer-chase for every `.val()` that happens on a `mat<var>`, and it's the pointer chasing that is slow. I can just ditch the `to_var_value` stuff (it's gonna be faster than no autodiff either way) but I thought I made a pretty good argument back here: https://github.com/stan-dev/math/pull/2164#discussion_r513022661.

Reply:
This is the big Q btw. Can do the control flow thing sure.